### PR TITLE
TW-334: Sticky timestamp inside chat

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket 

- #334 

### Resolved

https://github.com/linagora/twake-on-matrix/assets/99852347/cac0a5f0-762d-4fb9-b40c-24249129bb2f

https://github.com/linagora/twake-on-matrix/assets/99852347/03ce9375-7155-4026-b9ec-ded1f32a2d7e

### CODE REVIEW

**Comment by imGok** [comment]
The sticky timestamp does not stick perfectly to the chat timestamp, is there a way to improve the accuracy ?

**Comment by sherlockvn** [comment]
image
When i scroll to here, but the 26/1 is still show

File: lib/pages/chat/chat.dart [hunk 1]
```diff
@@ -1537,6 +1546,48 @@ class ChatController extends State<Chat>
...
+    }
+  }
+
+  void handleScrollEndNotification() async {
+    Logs().d('Chat::handleScrollEndNotification() - End of scroll');
+    if (PlatformInfos.isMobile) {
+      _currentChatScrollState = ChatScrollState.endScroll;
+      await Future.delayed(_delayHideStickyTimestampHeader);
+      _handleHideStickyTimestamp();
+    }
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1580]: why only mobile

Comment (nqhhdev) [reply lib/pages/chat/chat.dart:1580]: Only mobile, in web don't hide.

**Comment by nqhhdev** [comment]
> image When i scroll to here, but the 26/1 is still show

Yes, i know. I'm hiring new lib or new solutions for detect position of message. Because `visibility_detector` can't work well when we're scrolling very fast

**Comment by hoangdat** [comment]
please try to remove this kind of overlapsed

image

**Comment by hoangdat** [comment]
https://github.com/linagora/twake-on-matrix/assets/6462404/4ffac350-e430-4738-bcea-98bb0edf162f

File: lib/pages/chat/events/message/message.dart [hunk 2]
```diff
@@ -70,6 +71,7 @@ class Message extends StatelessWidget {
     this.onMenuAction,
     this.markedUnreadLocation,
     this.focusNode,
+    this.hideTimeStamp,
```
Comment (sherlockvn) [inline lib/pages/chat/events/message/message.dart:74]: add default value for `hideTimeStamp`

File: lib/pages/chat/sticky_timestamp_widget.dart [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    this.isStickyHeader = false,
+  }) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Align(
+      alignment: Alignment.topCenter,
+      child: UnconstrainedBox(
+        child: Opacity(
+          opacity: isStickyHeader ? 0.8 : 1.0,
```
Comment (sherlockvn) [inline lib/pages/chat/sticky_timestamp_widget.dart:20]: create style file

Comment (sherlockvn) [reply lib/pages/chat/sticky_timestamp_widget.dart:20]: or put it in style

File: lib/utils/extension/global_key_extension.dart [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter/material.dart';
+
+extension GlobalKeyExtension on GlobalKey {
+  Rect? get globalPaintBounds {
```
Comment (sherlockvn) [inline lib/utils/extension/global_key_extension.dart:4]: `globalPaintBoundsRect`

File: lib/pages/chat/chat.dart [hunk 5]
```diff
@@ -1537,6 +1550,60 @@ class ChatController extends State<Chat>
...
+  ) {
+    final stickyTimestampHeight =
+        stickyTimestampKey.globalPaintBounds?.height ?? 0;
+    return deltaTop < viewPortDimension - stickyTimestampHeight &&
+        deltaBottom > viewPortDimension - stickyTimestampHeight;
+  }
+
+  void _handleHideStickyTimestamp() {
+    Logs().d('Chat::_handleHideStickyTimestamp() - Hide sticky timestamp');
+    if (stickyTimestampNotifier.value != null) {
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1585]: no need the condition

File: lib/pages/chat/chat.dart [hunk 6]
```diff
@@ -1537,6 +1550,60 @@ class ChatController extends State<Chat>
...
+        stickyTimestampNotifier.value ??= dateTime;
+        if (stickyTimestampNotifier.value?.day != dateTime.day) {
+          Logs().d(
+            'Chat::handleDisplayStickyTimestamp() StickyTimestampNotifier - ${stickyTimestampNotifier.value}',
+          );
+          Logs().d(
+            'Chat::handleDisplayStickyTimestamp() CurrentDateTimeEvent - $_currentDateTimeEvent',
+          );
+          stickyTimestampNotifier.value = dateTime;
+        }
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1567]: put it in a method with clear what it does

**Comment by sherlockvn** [comment]
it works well from my side, including jumping from messages in search and pinned messages

**Comment by sherlockvn** [comment]
Please test in mobile when jump to messages, pinned messages also

File: lib/pages/chat/sticky_timestamp_widget.dart (hunk 3 above)
Comment (nqhhdev) [reply lib/pages/chat/sticky_timestamp_widget.dart:20]: IMO: It's a small widget, no need to create style file

Comment (sherlockvn) [reply lib/pages/chat/sticky_timestamp_widget.dart:20]: lol you just lazy

**Review by sherlockvn — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**